crates/jrsonnet-types/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+use std::fmt::Display;
+
+#[macro_export]
+macro_rules! ty {
+	([$inner:tt]) => {{
+		static VAL: &'static ComplexValType = &ty!($inner);
+		match VAL {
+			ComplexValType::Any => ComplexValType::Simple(ValType::Arr),
+			_ => ComplexValType::ArrayRef(&VAL),
+		}
+	}};
+	(bool) => {
+		ComplexValType::Simple(ValType::Bool)
+	};
+	(null) => {
+		ComplexValType::Simple(ValType::Null)
+	};
+	(str) => {
+		ComplexValType::Simple(ValType::Str)
+	};
+	(char) => {
+		ComplexValType::Char
+	};
+	(num) => {
+		ComplexValType::Simple(ValType::Num)
+	};
+	(number(($min:expr)..($max:expr))) => {{
+		ComplexValType::BoundedNumber($min, $max)
+	}};
+	(obj) => {
+		ComplexValType::Simple(ValType::Obj)
+	};
+	(any) => {
+		ComplexValType::Any
+	};
+	(fn.any) => {
+		ComplexValType::Simple(ValType::Func)
+	};
+	(($($a:tt) |+)) => {{
+		static CONTENTS: &'static [ComplexValType] = &[
+			$(ty!($a)),+
+		];
+		ComplexValType::UnionRef(CONTENTS)
+	}};
+	(($($a:tt) &+)) => {{
+		static CONTENTS: &'static [ComplexValType] = &[
+			$(ty!($a)),+
+		];
+		ComplexValType::SumRef(CONTENTS)
+	}};
+}
```

Comment:
Sholdn't absolute `$crate`-based paths be used for an exported macro?

Reply:
Yep, it should, but tgis is temporary syntax
I want to write proc macro instead for parsing typescript-like type annotations

Reply:
Proc macro would be too much, leaving it as is (With paths moved to $crate)